File: packages/serverless/lib/plugins/aws/bedrock-agentcore/examples/langgraph-basic-code/agent.py
"""
Minimal LangGraph agent with simple built-in tools.

This agent demonstrates:
- BedrockAgentCoreApp entrypoint pattern
- LangGraph with Claude Sonnet 4.5
- Simple tool integration (calculator, time)
- Code deployment (no Docker required)
"""

from typing import Annotated
from typing_extensions import TypedDict
from datetime import datetime

from langchain.chat_models import init_chat_model
from langchain_core.tools import tool
from langgraph.graph import StateGraph, START
from langgraph.graph.message import add_messages
from langgraph.prebuilt import ToolNode, tools_condition

from bedrock_agentcore.runtime import BedrockAgentCoreApp

# Initialize the AgentCore application
app = BedrockAgentCoreApp()

# Initialize Claude Sonnet 4.5 via US inference profile
llm = init_chat_model(
    "us.anthropic.claude-sonnet-4-5-20250929-v1:0",
    model_provider="bedrock_converse",
)

# Define simple tools using @tool decorator
@tool
def get_current_time(timezone: str = "UTC") -> str:
    """Get the current date and time."""
    return f"Current time: {datetime.now().strftime('%Y-%m-%d %H:%M:%S')} {timezone}"

@tool
def add(a: float, b: float) -> str:
    """Add two numbers together."""
    return f"Result: {a + b}"

@tool
def multiply(a: float, b: float) -> str:
    """Multiply two numbers together."""
    return f"Result: {a * b}"

tools = [get_current_time, add, multiply]
llm_with_tools = llm.bind_tools(tools)

# Define the conversation state
# LangGraph uses TypedDict to track state across nodes
class State(TypedDict):
    messages: Annotated[list, add_messages]


# Create the LangGraph
graph_builder = StateGraph(State)


def chatbot(state: State):
    """
    Main chatbot node that invokes the LLM.
    The LLM can decide to use tools or respond directly.
    """
    return {"messages": [llm_with_tools.invoke(state["messages"])]}


# Add nodes to the graph
graph_builder.add_node("chatbot", chatbot)
graph_builder.add_node("tools", ToolNode(tools=tools))

# Add edges
# After chatbot, conditionally go to tools if tool calls are requested
graph_builder.add_conditional_edges("chatbot", tools_condition)
# After using tools, return to chatbot to process results
graph_builder.add_edge("tools", "chatbot")
# Start at the chatbot node
graph_builder.add_edge(START, "chatbot")

# Compile the graph
graph = graph_builder.compile()


# Define the entrypoint for AgentCore
@app.entrypoint
def agent_invocation(payload, context):
    """
    AgentCore invokes this function with:
    - payload: dict with user input (typically contains 'prompt' key)
    - context: runtime context information (request ID, etc.)

    Returns:
    - dict with 'result' key containing the agent's response
    """
    user_message = payload.get("prompt", "Hello! How can I help you?")

    print(f"Received message: {user_message}")

    # Invoke the LangGraph with the user message
    result = graph.invoke({
        "messages": [{"role": "user", "content": user_message}]
    })

    # Extract the final message from the graph
    final_message = result['messages'][-1].content

    print(f"Responding with: {final_message}")

    # Return the agent's response
    return {
        "result": final_message
    }


# Start the AgentCore app
if __name__ == "__main__":
    app.run()

File: packages/serverless/lib/plugins/aws/bedrock-agentcore/examples/javascript/langgraph-basic-dockerfile/agent.js
/**
 * Minimal LangGraph JS agent with simple built-in tools.
 *
 * This agent demonstrates:
 * - BedrockAgentCoreApp entrypoint pattern for JavaScript
 * - LangChain createAgent (backed by LangGraph) with Claude Sonnet 4.5 via Bedrock
 * - Simple tool integration (calculator, time)
 * - Docker-based deployment
 */

import { BedrockAgentCoreApp } from 'bedrock-agentcore/runtime'
import { createAgent } from 'langchain'
import { ChatBedrockConverse } from '@langchain/aws'
import { tool } from '@langchain/core/tools'
import { z } from 'zod'

// Initialize Claude Sonnet 4.5 via US inference profile
const model = new ChatBedrockConverse({
  model: 'us.anthropic.claude-sonnet-4-5-20250929-v1:0',
  region: process.env.AWS_REGION || 'us-east-1',
})

// Define simple tools using the tool() helper

/**
 * Get the current date and time
 */
const getCurrentTime = tool(
  async ({ timezone }) => {
    const now = new Date()
    const options = {
      timeZone: timezone || 'UTC',
      dateStyle: 'full',
      timeStyle: 'long',
    }
    return `Current time: ${now.toLocaleString('en-US', options)}`
  },
  {
    name: 'get_current_time',
    description:
      'Get the current date and time. Optionally specify a timezone.',
    schema: z.object({
      timezone: z
        .string()
        .optional()
        .describe(
          'Timezone (e.g., "America/New_York", "Europe/London", "UTC")',
        ),
    }),
  },
)

/**
 * Add two numbers together
 */
const add = tool(
  async ({ a, b }) => {
    const result = a + b
    return `${a} + ${b} = ${result}`
  },
  {
    name: 'add',
    description: 'Add two numbers together.',
    schema: z.object({
      a: z.number().describe('First number'),
      b: z.number().describe('Second number'),
    }),
  },
)

/**
 * Multiply two numbers together
 */
const multiply = tool(
  async ({ a, b }) => {
    const result = a * b
    return `${a} × ${b} = ${result}`
  },
  {
    name: 'multiply',
    description: 'Multiply two numbers together.',
    schema: z.object({
      a: z.number().describe('First number'),
      b: z.number().describe('Second number'),
    }),
  },
)

/**
 * Divide two numbers
 */
const divide = tool(
  async ({ a, b }) => {
    if (b === 0) {
      return 'Error: Cannot divide by zero'
    }
    const result = a / b
    return `${a} ÷ ${b} = ${result}`
  },
  {
    name: 'divide',
    description: 'Divide two numbers.',
    schema: z.object({
      a: z.number().describe('Dividend (number to divide)'),
      b: z.number().describe('Divisor (number to divide by)'),
    }),
  },
)

// Collect all tools
const tools = [getCurrentTime, add, multiply, divide]

// Create LangGraph agent
// This creates a graph that alternates between calling the LLM and executing tools
const agent = createAgent({
  model,
  tools,
})

// Initialize the AgentCore application with invocation handler
const app = new BedrockAgentCoreApp({
  invocationHandler: {
    // Define the expected request schema
    requestSchema: z.object({
      prompt: z.string().describe('The user message to process'),
    }),

    // Process incoming requests (non-streaming)
    async process(request, context) {
      console.log(`Received message: ${request.prompt}`)
      console.log(`Request ID: ${context?.requestId || 'unknown'}`)

      // Invoke the LangGraph agent with the user message
      const result = await agent.invoke({
        messages: [{ role: 'user', content: request.prompt }],
      })

      // Extract the final message from the graph result
      const finalMessage = result.messages[result.messages.length - 1]
      const response = finalMessage.content

      console.log(`Responding with: ${response}`)

      // Return the response (non-streaming)
      return response
    },
  },
})

// Start the AgentCore application
// This starts an HTTP server on port 8080 (default)
app.run()
